pysb/generator/kappa.py

```diff
@@ -216,22 +271,82 @@ def format_site_condition(site, state):
     # Something else
     else:
         raise KappaException("Kappa generator has encountered an unknown "
-                        "element in a rule pattern site condition.")
+                             "element in a rule pattern site condition.")
     return '%s%s' % (site, state_code)
 
-def expression_to_muparser(expression):
-    """Render the Expression as a Kappa compatible string."""
-    # sympy.printing.sstr is the preferred way to render an Expression as a
-    # string (rather than, e.g., str(Expression.expr) or repr(Expression.expr).
-    # Note: "For large expressions where speed is a concern, use the setting
-    # order='none'"
-    code = sympy.printing.sstr(expression.expr, order='none')
-    code = code.replace('\n @', '')
-    code = code.replace('**', '^')
-    # kasim syntax cannot handle Fortran scientific notation (must use 'e'
-    # instead of 'd')
-    code = sub('(?<=[0-9])d', 'e', code)
-    return code
+
+class KappaPrinter(StrPrinter):
+    def __init__(self, **settings):
+        super(KappaPrinter, self).__init__(settings)
+
+    def _print_Piecewise(self, expr):
+        if expr.args[-1][1] is not sympy.true:
+            raise NotImplementedError('Piecewise statements are only '
+                                      'supported if convertible to Kappa '
+                                      'ternary statements')
+
+        if_stmt = self._print(expr.args[-1][0])
+        for pos in range(len(expr.args) - 2, -1, -1):
+            if_stmt = '{} [?] {} [:] {}'.format(
+                self._print(expr.args[pos][1]),
+                self._print(expr.args[pos][0]),
+                if_stmt
+            )
+
+        return if_stmt
+
+    def _print_Pow(self, expr, rational=False):
+        """ Adapted from sympy/printing/str.py """
+        if expr.exp is S.Half and not rational:
+            return "[sqrt](%s)" % self._print(expr.base)
+
+        if expr.is_commutative:
+            if -expr.exp is S.Half and not rational:
+                # Note: Don't test "expr.exp == -S.Half" here, because that
+                # will match -0.5, which we don't want.
+                return "1/[sqrt](%s)" % self._print(expr.base)
+
+        return super(KappaPrinter, self)._print_Pow(expr, rational)\
+            .replace('**', '^')
+
+    def _print_Pi(self, expr):
+        return '[pi]'
```

Comment:
I guess `pi` is special-cased in Kappa? There are a few other built-in constants like that in sympy but maybe the only useful one in this context is `E`. Is that special cased in Kappa? If not, it could be implemented here anyway as `return str(float(expr))`.

Reply:
Yes, `[pi]` is lower case in Kappa. It appears `E` works as `[E]`, but is undocumented. I implemented it as such.